### PR TITLE
Add realtime feedback modal

### PR DESCRIPTION
**Pull request by dem4ron**
This PR adds a modal after submitting, looks for automated feedback, and shows them if those are present.

Latest feedback look (july 28):
Screenshot 2023-07-28 at 11 37 10

Screenshot 2023-05-10 at 15 12 21

Screenshot 2023-03-30 at 16 38 39

Edit: the one below is an older printscreen, but the only change is the button says "Continue anyway" instead of "Continue"

Screenshot 2023-03-23 at 17 34 07

Screenshot 2023-03-23 at 17 33 47

Screenshot 2023-03-24 at 14 45 23

celebratory feedback:

Screenshot 2023-03-30 at 12 09 31

TODO:
- [x] Refactor
- [x] Write tests
- [x] Handle celebratory feedback

### CODE REVIEW

File: app/javascript/components/modals/realtime-feedback-modal/CheckingForAutomatedFeedback.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  )
+}
+
+function TakingTooLong(): JSX.Element {
+  return (
+    <div>
+      <p className="text-p-base">
+        Sorry, this is taking a little long.
+        <br />
+        We&apos;ll continue generating feedback in the background.
```
Comment (ErikSchierboom) [inline app/javascript/components/modals/realtime-feedback-modal/CheckingForAutomatedFeedback.tsx]: Maybe change this to:

```suggestion
        We&apos;ll continue gathering feedback in the background.
```

File: app/javascript/components/student/IterationsList.tsx [hunk 2]
```diff
@@ -109,9 +109,7 @@ export const IterationsList = ({
     }
 
     if (isOpen.length === 0) {
-      setIsOpen(
-        resolvedData.iterations.map((iteration, i) => (i === 0 ? true : false))
-      )
+      setIsOpen(resolvedData.iterations.map((_, i) => (i === 0 ? true : false)))
```
Comment (ErikSchierboom) [inline app/javascript/components/student/IterationsList.tsx]: This should be the same:

```suggestion
      setIsOpen(resolvedData.iterations.map((_, i) => i === 0)))
```

File: test/system/flows/edit_solution_test.rb [hunk 3]
```diff
@@ -119,11 +125,207 @@ class EditSolutionTest < ApplicationSystemTestCase
...
+          Submission::TestRunsChannel.broadcast!(test_run)
+          assert_button "Submit", disabled: false
+
+          visit edit_track_exercise_path(solution.track, solution.exercise)
+          within(".lhs-footer") { click_on "Submit" }
+
+          assert_text "Checking for automated feedback..."
+          assert_text "Continue anyway"
+          sleep(10)
+          assert_text "Sorry, this is taking a little long."
```
Comment (ErikSchierboom) [inline test/system/flows/edit_solution_test.rb]: What happens when the wait timeout is exceeded? As in: what is shown then?

File: test/system/flows/edit_solution_test.rb [hunk 4]
```diff
@@ -119,11 +125,207 @@ class EditSolutionTest < ApplicationSystemTestCase
...
+          submission.update!(representation_status: :generated, analysis_status: :completed)
+          solution.reload
+          Submission::TestRunsChannel.broadcast!(test_run)
+          SolutionChannel.broadcast!(solution)
+          assert_text "Essential"
+          assert_text "We've found some automated feedback"
+        end
+      end
+
+      test "user reads essential feedback, goes back to exercise, revisits cached feedback by pressing submit button" do
```
Comment (ErikSchierboom) [inline test/system/flows/edit_solution_test.rb:334]: Is there also a test that combined analyzer and representer feedback?

**Review by ErikSchierboom — COMMENTED** [review]
Great work!

File: app/javascript/components/modals/realtime-feedback-modal/feedback-content/FoundAutomatedFeedback.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import React from 'react'
+import { AnalyzerFeedback } from '@/components/student/iterations-list/AnalyzerFeedback'
+import { RepresenterFeedback } from '@/components/student/iterations-list/RepresenterFeedback'
+import { AnalysisStatusSummary } from '@/components/track/iteration-summary/AnalysisStatusSummary'
+import { GoBackToExercise, ContinueButton } from './FeedbackContentButtons'
+import { FeedbackContentProps } from '../FeedbackContent'
+
+const HEADLINE = [
+  "We've found some automated feedback",
+  "We've found a celebratory automated feedback! 🎉",
```
Comment (ErikSchierboom) [inline app/javascript/components/modals/realtime-feedback-modal/feedback-content/FoundAutomatedFeedback.tsx]:
```suggestion
  "We've found celebratory automated feedback! 🎉",
```

File: test/system/flows/edit_solution_test.rb [hunk 6]
```diff
@@ -119,11 +125,207 @@ class EditSolutionTest < ApplicationSystemTestCase
...
+
+          visit edit_track_exercise_path(solution.track, solution.exercise)
+          within(".lhs-footer") { click_on "Submit" }
+          assert_text "Checking for automated feedback"
+          sleep(1)
+          submission.update!(representation_status: :generated, analysis_status: :completed)
+          solution.reload
+          Submission::TestRunsChannel.broadcast!(test_run)
+          SolutionChannel.broadcast!(solution)
+          assert_text "We've found a celebratory automated feedback! 🎉"
```
Comment (ErikSchierboom) [inline test/system/flows/edit_solution_test.rb]:
```suggestion
          assert_text "We've found celebratory automated feedback! 🎉"
```

File: app/javascript/components/student/IterationsList.tsx (hunk 2 above)
Comment (dem4ron) [reply app/javascript/components/student/IterationsList.tsx]: I probably auto-formatted this older bit. thanks for catching it!

File: app/css/components/details.css [hunk 7]
```diff
@@ -13,6 +13,7 @@
                 height: 12px;
                 width: 12px;
                 @apply ml-16 filter-textColor6;
+
```
Comment (iHiD) [inline app/css/components/details.css:16]:
```suggestion
```

File: app/helpers/react_components/editor.rb [hunk 8]
```diff
@@ -37,27 +37,47 @@ def data
...
+        track: {
+          title: track.title,
+          slug: track.slug,
+          icon_url: track.icon_url,
+          median_wait_time: track.median_wait_time
+        }
+      }
+    end
+
+    # TODO: clean this up, and maybe enough to get the latest iteration?
```
Comment (iHiD) [inline app/helpers/react_components/editor.rb:72]: Do you want to act on this todo?

File: app/helpers/react_components/editor.rb [hunk 9]
```diff
@@ -96,6 +120,11 @@ def debugging_instructions
       Markdown::Parse.(track.debugging_instructions)
     end
 
+    memoize
+    def user_track
+      UserTrack.for(solution.user, track)
+    end
```
Comment (iHiD) [inline app/helpers/react_components/editor.rb:126]:
```suggestion
    def user_track = UserTrack.for(solution.user, track)
```

File: app/javascript/components/Editor.tsx [hunk 10]
```diff
@@ -181,20 +203,27 @@ export default ({
...
 
-    dispatch({ status: EditorStatus.CREATING_ITERATION })
-
-    createIteration(submission, {
-      onSuccess: async (iteration) => {
-        await cache.invalidateQueries([getCacheKey(track.slug, exercise.slug)])
-        redirectTo(iteration.links.solution)
-      },
-    })
+    showFeedbackModal()
```
Comment (iHiD) [inline app/javascript/components/Editor.tsx]:
```suggestion
    showFeedbackModal()
    
```

File: app/javascript/components/Editor.tsx [hunk 11]
```diff
@@ -521,6 +550,20 @@ export default ({
...
+            discussion={discussion}
+            mentoringStatus={mentoringStatus}
+            onSubmit={submit}
+            solution={solution}
+            track={track}
+            request={request}
+            submission={submission}
+            exercise={exercise}
+            trackObjectives={trackObjectives}
+            links={{ ...links, redirectToExerciseLink: redirectLink }}
```
Comment (iHiD) [inline app/javascript/components/Editor.tsx:573]:
```suggestion
            links={{ ...links, redirectToExercise: redirectLink }}
```

This is more of a general comment: I feel like if we have `links: { ... }` then the things inside `...` probably don't need the `Link` suffix.

File: app/javascript/components/editor/ChatGptFeedback/AskChatGpt.tsx [hunk 12]
```diff
@@ -65,14 +66,7 @@ function AskingChatGpt() {
...
       <GraphicalIcon icon="spinner" className="animate-spin-slow" />
-      <div className="progress">
-        <div
-          className="bar"
-          style={{
-            animationDuration: '15s',
-          }}
-        />
-      </div>
+      <LoadingBar animationDuration={15} />
```
Comment (iHiD) [inline app/javascript/components/editor/ChatGptFeedback/AskChatGpt.tsx:69]: Nice.

File: app/javascript/components/Editor.tsx (hunk 11 above)
Comment (dem4ron) [reply app/javascript/components/Editor.tsx:573]: I usually add it, if only one link is passed down. and I want to know it is a link

File: app/javascript/components/modals/realtime-feedback-modal/feedback-content/found-automated-feedback/index.tsx [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+import {
+  GoBackToExercise,
+  ContinueButton,
+} from '../../components/FeedbackContentButtons'
+import { FeedbackContentProps } from '../../FeedbackContent'
+import { FooterButtonContainer } from '../../components'
+
+const HEADLINE = [
+  "Here's a suggestion on how to improve your code…",
+  "We've found celebratory automated feedback! 🎉",
```
Comment (iHiD) [inline app/javascript/components/modals/realtime-feedback-modal/feedback-content/found-automated-feedback/index.tsx]:
```suggestion
  "We've got some positive feedback for you! 🎉",
```

File: app/javascript/components/modals/realtime-feedback-modal/feedback-content/found-automated-feedback/index.tsx [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+import { AnalyzerFeedback } from './AnalyzerFeedback'
+import { RepresenterFeedback } from './RepresenterFeedback'
+import {
+  GoBackToExercise,
+  ContinueButton,
+} from '../../components/FeedbackContentButtons'
+import { FeedbackContentProps } from '../../FeedbackContent'
+import { FooterButtonContainer } from '../../components'
+
+const HEADLINE = [
```
Comment (iHiD) [inline app/javascript/components/modals/realtime-feedback-modal/feedback-content/found-automated-feedback/index.tsx]:
```suggestion
const HEADLINES = [
```

File: app/javascript/components/modals/realtime-feedback-modal/feedback-content/found-automated-feedback/index.tsx [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+}: Pick<
+  FeedbackContentProps,
+  'latestIteration' | 'track' | 'onClose' | 'links'
+> & { onContinue: () => void; celebratory?: boolean }): JSX.Element {
+  return (
+    <>
+      <div className="flex gap-40 items-start">
+        <div className="flex-col items-left">
+          <div className="text-h4 mb-16 flex c-iteration-summary">
+            {HEADLINE[+celebratory]}
```
Comment (iHiD) [inline app/javascript/components/modals/realtime-feedback-modal/feedback-content/found-automated-feedback/index.tsx]: What's this `+` do?

Comment (dem4ron) [reply app/javascript/components/modals/realtime-feedback-modal/feedback-content/found-automated-feedback/index.tsx]: it casts other types into number. in this case into 1 or 0 to use it as an array index (but this code is deprecated now)

**Pull request merged**